File: android/app/src/test/kotlin/app/codexlauncher/capability/reply/HonestReplyWordTest.kt
package app.codexlauncher.capability.reply

import app.codexlauncher.capability.notifications.ProbeVerdict
import app.codexlauncher.capability.notifications.ReplySource
import app.codexlauncher.capability.reply.guard.ReplyCap
import app.codexlauncher.capability.reply.guard.ReplyGuard
import app.codexlauncher.capability.reply.request.DeviceReplyRequest
import app.codexlauncher.capability.reply.request.ReplyHandleSource
import app.codexlauncher.connection.protocol.MessageType
import app.codexlauncher.connection.protocol.ProtocolCodec
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The phone half of "stop calling it delivered".
 *
 * Everything this phone observes when it fires a reply is one line:
 * `actionIntent.send(context, 0, fillInIntent)` did not throw
 * (`AndroidReplyDispatch.kt:76`). That proves Android handed the text to the
 * app that posted the notification. It does not prove WhatsApp sent anything,
 * and it certainly does not prove anybody received anything — the app may be
 * queued behind no signal, logged out, or may simply drop it.
 *
 * `DELIVERED` and `"delivered"` both claim otherwise, and this whole route
 * exists because the route it replaced claimed "sent" without checking.
 *
 * The word becomes `handed_to_the_app` — the largest true statement available.
 * It is not a hedge: three of the four outcomes stay exactly as certain as they
 * were, and the tests below hold them there, because "we can't be sure" is its
 * own dishonesty when we can. Somebody told nothing was sent can go and send
 * it; somebody told "maybe" can do nothing at all.
 *
 * The Mac's half of this is
 * `companion/internal/app/mobilesession/honest_reply_word_test.go`, and the two
 * must agree word for word — a word one machine does not know does not arrive
 * smaller, it makes the whole frame get dropped and the person is told nothing.
 */
class HonestReplyWordTest {

    private val firedWord = "handed_to_the_app"
    private val text = "on my way"

    private fun handle(
        verdict: ProbeVerdict = ProbeVerdict.CAN_REPLY,
        notificationLive: Boolean = true,
    ) = ReplyHandle(
        packageName = "com.whatsapp",
        appLabel = "WhatsApp",
        verdict = verdict,
        remoteInputKey = "android.intent.extra.text",
        source = ReplySource.SHADE,
        notificationLive = notificationLive,
        conversationKey = "0|com.whatsapp|1|null|10",
    )

    private fun boxesFor(vararg handles: ReplyHandle) = ReplyHandleSource { _ -> handles.toList() }

    private fun dispatcher(result: DeliveryResult) =
        object : ReplyDispatch {
            override fun deliver(conversationKey: String, remoteInputKey: String, text: String) = result
        }

    private fun carryOut(boxes: ReplyHandleSource, dispatch: ReplyDispatch?) =
        DeviceReplyRequest.carryOut(
            handle = "maya",
            text = text,
            boxes = boxes,
            dispatch = dispatch,
            // Fresh per call, so no stop or send cap can ever decide a test in
            // this file. What is under test here is the honesty of the word,
            // not the guard — those rules live in `guard/GuardedReplyTest.kt`.
            guard = ReplyGuard(now = { 0L }, cap = ReplyCap.shipped),
        )

    private fun resultFrame(outcome: String) =
        """{"version":{"major":1,"minor":0},"messageId":"dev-r-1","sender":"phone","type":"device_action_result","body":{"requestId":"cap-action-1","outcome":"$outcome"}}"""

    /** The headline: what the phone says after a send that did not throw. */
    @Test
    fun `a reply the app accepted is not reported as delivered`() {
        val fired = dispatcher(DeliveryResult.entries.first { it.name != "NOTIFICATION_GONE" && it.name != "FAILED" })
        assertEquals(firedWord, carryOut(boxesFor(handle()), fired))
    }

    /**
     * The claim has to be gone from the source too, not merely unused. A value
     * still named DELIVERED is a standing invitation for the next person to
     * write the sentence it implies.
     */
    @Test
    fun `nothing in the send path is still named delivered`() {
        val names = DeliveryResult.entries.map { it.name }
        assertTrue(
            "DeliveryResult still has a value claiming delivery: $names",
            names.none { it.contains("DELIVER") },
        )
    }

    /** Both machines have to know the same four words. */
    @Test
    fun `the wire carries the new word and no longer carries the old one`() {
        assertEquals(MessageType.DEVICE_ACTION_RESULT, ProtocolCodec.decodeText(resultFrame(firedWord)).type)
        for (outcome in listOf("delivered", "sent", "HANDED_TO_THE_APP", "handed")) {
            assertThrows(Exception::class.java) { ProtocolCodec.decodeText(resultFrame(outcome)) }
        }
    }

    /**
     * The control that stops this becoming a blanket hedge. These three know
     * for certain that nothing was sent, and they must keep saying so.
     */
    @Test
    fun `the outcomes that know nothing was sent are untouched`() {
        assertEquals("notification_gone", carryOut(ReplyHandleSource { emptyList() }, dispatcher(DeliveryResult.FAILED)))
        assertEquals("notification_gone", carryOut(boxesFor(handle()), dispatcher(DeliveryResult.NOTIFICATION_GONE)))
        assertEquals("failed", carryOut(boxesFor(handle()), dispatcher(DeliveryResult.FAILED)))
        assertEquals("refused", carryOut(boxesFor(handle()), dispatch = null))
        assertEquals("refused", carryOut(boxesFor(handle(verdict = ProbeVerdict.NO_REPLY_BOX)), dispatcher(DeliveryResult.FAILED)))
    }

    /**
     * Second control. Renaming one word in a set of four is exactly the change
     * that leaves a fifth word loose, so every word this path can produce is
     * checked against the wire's set rather than against a list written here.
     */
    @Test
    fun `every word this path can produce is one the wire accepts`() {
        val fired = dispatcher(DeliveryResult.entries.first { it.name != "NOTIFICATION_GONE" && it.name != "FAILED" })
        val produced = listOf(
            carryOut(boxesFor(handle()), fired),
            carryOut(boxesFor(handle(notificationLive = false)), fired),
            carryOut(boxesFor(handle(verdict = ProbeVerdict.NO_REPLY_BOX)), fired),
            carryOut(boxesFor(handle()), dispatcher(DeliveryResult.FAILED)),
            carryOut(boxesFor(handle()), dispatch = null),
            carryOut(ReplyHandleSource { emptyList() }, fired),
        )
        for (word in produced.toSet()) {
            assertEquals(
                "the phone produced $word, which the Mac would drop the whole frame over",
                MessageType.DEVICE_ACTION_RESULT,
                ProtocolCodec.decodeText(resultFrame(word)).type,
            )
        }
    }
}

File: companion/internal/app/mobilesession/honest_reply_word_test.go
package mobilesession

import (
	"context"
	"strings"
	"testing"

	"github.com/codex-launcher/codex-launcher/companion/internal/capability/manifest"
	"github.com/codex-launcher/codex-launcher/companion/internal/mobileapi/contract"
)

// The word the phone sends back when a reply is fired is `delivered`, and it
// is not true.
//
// Here is everything the phone actually observes. `AndroidReplyDispatch.deliver`
// puts the text into the notification action's RemoteInput bundle and calls
// `actionIntent.send(...)`. If that call does not throw, it returns DELIVERED.
// That is the whole basis for the claim. What the call proves is that Android
// handed the bundle to the messaging app that posted the notification. WhatsApp
// never told us anything. The recipient's phone never told us anything. The
// message may be sitting in a send queue behind no signal; the app may have
// dropped it; the account may be logged out.
//
// This plan exists because the route it replaced claimed "sent" without
// checking. Repeating that under a friendlier word is the same lie with better
// manners, and it is worse here than it was there, because the person reading
// "Reply sent." has no reason to open the app and look.
//
// Two things this is deliberately NOT:
//
//   - **It is not a lower ceiling.** A ceiling answers "how much is left for
//     the user to do", and once the text has gone into the reply box there is
//     nothing left for them to do. `completes` is right and stays. The doubt
//     here is about certainty, not effort, and this codebase already keeps
//     those two apart on purpose.
//
//   - **It is not `done: false`.** Same reason. The work was carried out. What
//     is unproven is the result, and reporting the work as unfinished would
//     put the person back in front of a task they cannot do anything about.
//
// So the change is to the word, and to the sentence the person reads. The word
// becomes `handed_to_the_app`, which is the largest true statement available:
// we handed the text to the app that owns the conversation, and that is where
// what we can see ends.
//
// The controls below matter more than the headline. "Never claim anything"
// passes the first test and destroys the other three outcomes, which really do
// know that nothing was sent.

const firedWord = "handed_to_the_app"

// The headline. The word the wire carries for a fired reply must not be a
// claim about arrival.
func TestTheWordForAFiredReplyDoesNotClaimItArrived(t *testing.T) {
	handler, sender, _ := newDeviceWorkHandler(t)
	handOffDeclaring(t, handler, sender, manifest.Completes)

	if err := handler.Handle(context.Background(), sender, decode(t, answerFrame(firedWord))); err != nil {
		t.Fatalf("the phone reported %q and the Mac would not accept it: %v", firedWord, err)
	}
	body := bodyOf(t, awaitSentMessage(t, sender.sent))

	if done, _ := body["done"].(bool); !done {
		t.Fatalf("a fired reply was reported as unfinished: %v — the doubt belongs in the word, not in the done flag", body)
	}
	if ceiling, _ := body["ceiling"].(string); ceiling != string(manifest.Completes) {
		t.Fatalf("ceiling = %q, want completes — a ceiling is about effort left, and none is", ceiling)
	}
}

// The old word has to actually be gone. Leaving it accepted "for compatibility"
// would mean a phone still shipping it is still making the claim, and nothing
// would ever surface that.
func TestTheOldWordIsGoneFromTheWire(t *testing.T) {
	if _, err := contract.DecodeText([]byte(answerFrame("delivered"))); err == nil {
		t.Fatal("the wire still accepts \"delivered\"; a word that is not true must not be reachable, not merely unused")
	}
}

// What the person reads. The frame's own word is invisible to them; this
// sentence is the whole of what they are told.
func TestWhatThePersonReadsDoesNotClaimTheMessageArrived(t *testing.T) {
	handler, sender, _ := newDeviceWorkHandler(t)
	handOffDeclaring(t, handler, sender, manifest.Completes)

	if err := handler.Handle(context.Background(), sender, decode(t, answerFrame(firedWord))); err != nil {
		t.Fatal(err)
	}
	body := bodyOf(t, awaitSentMessage(t, sender.sent))
	detail, _ := body["detail"].(string)

	if detail == "" {
		t.Fatal("the person is told nothing at all about a reply that was fired at a real conversation")
	}
	// "sent", "delivered" and "received" are all claims about something on the
	// other side of an app we cannot see into.
	for _, claim := range []string{"sent", "delivered", "received", "went through"} {
		if strings.Contains(strings.ToLower(detail), claim) {
			t.Fatalf("the person is told %q, which claims %q — we only know Android handed the text to the app", detail, claim)
		}
	}
}

// First control, and the one that stops the over-correction. Three of the four
// outcomes are certain, and softening them into "we can't tell" would be its
// own dishonesty — worse, because a person who is told nothing was sent can
// act, and a person told "maybe" cannot.
func TestTheOutcomesThatKnowNothingWasSentStillSaySo(t *testing.T) {
	for _, outcome := range []string{"notification_gone", "refused", "failed"} {
		handler, sender, _ := newDeviceWorkHandler(t)
		handOffDeclaring(t, handler, sender, manifest.Completes)

		if err := handler.Handle(context.Background(), sender, decode(t, answerFrame(outcome))); err != nil {
			t.Fatalf("%s: %v", outcome, err)
		}
		body := bodyOf(t, awaitSentMessage(t, sender.sent))

		if done, _ := body["done"].(bool); done {
			t.Fatalf("%s was reported as done: %v", outcome, body)
		}
		if detail, _ := body["detail"].(string); detail == "" {
			t.Fatalf("%s left the person with no explanation: %v", outcome, body)
		}
	}
}

// Second control. Renaming one word in a set of four is exactly the change that
// leaves the two machines disagreeing, and a word one side does not know is not
// a smaller message — the whole frame is dropped and the person sees nothing.
func TestTheWireKnowsFourWordsAndNoOthers(t *testing.T) {
	for _, outcome := range []string{firedWord, "notification_gone", "failed", "refused"} {
		if _, err := contract.DecodeText([]byte(answerFrame(outcome))); err != nil {
			t.Fatalf("the wire refused %q, which the phone can produce: %v", outcome, err)
		}
	}
	for _, outcome := range []string{"", "sent", "ok", "unknown", "HANDED_TO_THE_APP", "handed"} {
		if _, err := contract.DecodeText([]byte(answerFrame(outcome))); err == nil {
			t.Fatalf("the wire accepted %q, which nothing produces", outcome)
		}
	}
}
